File: tessoku/tessoku-book/src/bin/a08.rs
use proconio::input;

fn main() {
    input! {
      (h, w): (usize, usize),  // H,W <= 1500
      x: [[usize; w]; h],
      q: usize, // Q <= 10^5
      abcd: [(usize, usize, usize, usize); q]
    }

    // grid を 1-index にする
    let mut grid = vec![vec![0; w + 2]; h + 2];
    for i in 0..h {
        for j in 0..w {
            grid[i + 1][j + 1] = x[i][j];
        }
    }

    // O(Q)
    for &(a, b, c, d) in &abcd {
        // 縦横2回の累積和 O(HW)
        let mut sum = grid.clone();
        // h 方向の累積和
        for i in a..=c {
            for j in 1..=w {
                sum[i][j] += sum[i - 1][j];
            }
        }
        // w 方向の累積和
        for j in b..=d {
            for i in 1..=h {
                sum[i][j] += sum[i][j - 1];
            }
        }
        // abcd領域の合計
        let ans = sum[a - 1][b - 1] + sum[c][d] - sum[a - 1][d] - sum[c][b - 1];
        println!("{}", ans);
    }
}

File: tessoku/tessoku-book/src/a08.py
# [A08 - Two Dimensional Sum](https://atcoder.jp/contests/tessoku-book/tasks/tessoku_book_h)
def solve(H: int, W: int, X: list[list[int]], _Q: int, ABCD: list[list[int]]):
    # X: 0-indexed, S: 1-indexed, ABCD: 1-indexed
    # いもす法で累積和を計算する
    S = [[0] * (W + 2) for _ in range(H + 2)]
    # 縦
    for i in range(1, H + 1):
        for j in range(1, W + 1):
            S[i][j] += S[i - 1][j] + X[i][j]
    # 横
    for j in range(1, W + 1):
        for i in range(1, H + 1):
            S[i][j] += S[i][j - 1]
    # 左上+右下 - (左下+右上)
    for a, b, c, d in ABCD:
        ans = S[a - 1][b - 1] + S[c][d] - S[a - 1][d] - S[c][b - 1]
        print(ans)


if __name__ == "__main__":
    [H, W] = list(map(int, input().split()))
    X = [list(map(int, input().split())) for _ in range(H)]
    Q = int(input())
    ABCD = [list(map(int, input().split())) for _ in range(Q)]
    # X だけ 0-indexed だと不便なので 1-indexed に変換する
    X1 = [[0] * (W + 2) for _ in range(H + 2)]
    for i in range(H):
        for j in range(W):
            X1[i + 1][j + 1] = X[i][j]

    solve(H, W, X1, Q, ABCD)
